# Data Design & Data Wrangling

'Data wrangling' is the mundane labor of collecting, loading, transforming, and preparing data for exploration and analysis. Although it is hard to put a number on it, [estimates](https://www.nytimes.com/2014/08/18/technology/for-big-data-scientists-hurdle-to-insights-is-janitor-work.html) indicate that data scientists spend from 50 per cent to 80 per cent of their time on data wrangling. The importance of data wrangling has been recognised in [social sciencies](https://doi.org/10.1177/2378023118817378) as well. 

Although data wrangling may seem a-theoretical, some of the decisions you would make would in fact depend on how the data were generated or designed.

In this lab, we will first discuss issues related to data design and will then practice hands-on data wrangling with `pandas` and other Python libraries for data analysis. We will wrangle open data sets, including the Google Covid-19 Community Mobility Reports and Apple Mobility Trends Reports.

## Key themes
* What are data? How are data generated (i.e., data design)? 
* Data wrangling with `pandas`: loading, selecting, and transforming data.

## Learning resources

<i class="fas fa-book"></i> Matthew Salganik. [2.3 Ten common characteristics of big data](https://www.bitbybitbook.com/en/1st-ed/observing-behavior/). In _Bit by Bit: Social Research in the Digital Age_. [[Online version freely available]]((https://www.bitbybitbook.com/en/1st-ed/preface/))

<i class="fas fa-book"></i> Lisa Gitelman and Virginia Jackson. [“Raw Data” Is an Oxymoron. Introduction.](https://raley.english.ucsb.edu/wp-content/Engl800/RawData-excerpts.pdf) MIT Press.

<i class="fas fa-scroll"></i> Sabina Leonelli. [Scientific Research and Big Data.](https://plato.stanford.edu/cgi-bin/encyclopedia/archinfo.cgi?entry=science-big-data) The Stanford Encyclopedia of Philosophy.

<i class="fas fa-book"></i> <i class="fas fa-code"></i> Sam Lau, Joey Gonzalez, and Deb Nolan. [The Data Science Lifecycle](https://www.textbook.ds100.org/ch/01/lifecycle_intro.html). In [_Principles and Techniques of Data Science_](https://www.textbook.ds100.org/intro.html).

<i class="fas fa-scroll"></i> Steve Lohr. [For Big-Data Scientists, ‘Janitor Work’ Is Key Hurdle to Insights](https://www.nytimes.com/2014/08/18/technology/for-big-data-scientists-hurdle-to-insights-is-janitor-work.html). New York Times. 

<i class="fas fa-scroll"></i> Alexander Kindel et al. [Improving Metadata Infrastructure for Complex Surveys: Insights from the Fragile Families Challenge](https://doi.org/10.1177/2378023118817378). 
Socius.  

<i class="fas fa-play-circle"></i> <i class="fas fa-code"></i> [Introduction to Data Processing in Python with Pandas](https://www.youtube.com/watch?v=5rNu16O3YNE), SciPy 2019 Tutorial by Daniel Chen. See also freely available Jupyter notebooks: [01-intro.ipynb](https://github.com/valdanchev/scipy-2019-pandas/blob/master/notebooks/01-intro.ipynb) and [02-tidy.ipynb](https://github.com/valdanchev/scipy-2019-pandas/blob/master/notebooks/02-tidy.ipynb).

<br>


```{admonition} Activity: Discussing data biases
After reading the section [Ten common characteristics of big data](https://www.bitbybitbook.com/en/1st-ed/observing-behavior/) in Matthew Salganik's book _Bit by Bit_, we will discuss how each characteristic of big data applies to the digital data sets we are analysing in this and next sections, including the [Google COVID-19 Community Mobility Reports](https://www.google.com/covid19/mobility/) and [Apple Mobility Trends Reports](https://covid19.apple.com/mobility).
```

<img src="https://upload.wikimedia.org/wikipedia/commons/e/ed/Pandas_logo.svg" title='Pandas Logo' width="200" height="100" align="right" />

# Get started with [`pandas`](https://pandas.pydata.org/docs/getting_started/index.html#getting-started)
The `pandas` library:
* is a fast, powerful, and flexible open source tool for doing real world data analysis in Python.
* offers a diverse range of high-performance tools for data loading, cleaning, wrangling, merging, reshaping,  and summarising.
* is the go-to data science library in Python.

# Importing `pandas`

To use `pandas`, we first import the library via the Python's `import` command and give it the alias pd:

In [68]:
import pandas as pd
pd.__version__

'1.1.3'

# Loading your data

Pandas supports many data file formats, including csv, excel, sql, json.
For details, see [How do I read and write tabular data?](https://pandas.pydata.org/docs/getting_started/intro_tutorials/02_read_write.html#min-tut-02-read-write) For example, once you import pandas as pd, you can load a .csv file using the function `pd.read_csv()`, an Excel file using the function `pd.read_excel()`, and a JSON file using the function `pd.read_json()`.

<img src="https://pandas.pydata.org/docs/_images/02_io_readwrite.svg" width="800" height="400" >

# Loading Covid-19 data sets via URL 


<img src="https://www.google.com/covid19/static/reports-icon-grid.png" title='Google Covid-19 Community Mobility Data' width="200" height="100" align="right"/>

What are the Google Covid-19 [Community Mobility Reports](https://www.google.com/covid19/mobility/)?

* Aggregated and anonymized sets of freely available data that protect individual privacy.
* Shows trends of human mobility over time by country and region, across six categories of places, including retail and recreation, groceries and pharmacies, parks, transit stations, workplaces, and residential.
* The data shows how visitors to (or time spent in) categorized places change compared to baseline days. A baseline day represents a normal value for that day of the week. The baseline day is the median value from the 5‑week period Jan 3 – Feb 6, 2020. 

An overview of the data from the mobility data can be found [here](https://www.google.com/covid19/mobility/).

The code below loads the most recent version of the Covid-19 Google Community Mobility Reports as a `pandas` DataFrame object and assign the object to a variable called 'mobility_trends'

In [69]:
mobility_trends = pd.read_csv('https://www.gstatic.com/covid19/mobility/Global_Mobility_Report.csv')

/Users/valentindanchev/opt/anaconda3/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3146: DtypeWarning: Columns (4,5) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


`Pandas` load the data as a DataFrame object. We can confirm this by using the built-in function `type()`:

In [5]:
type(mobility_trends)

pandas.core.frame.DataFrame

## What is a [DataFrame](https://pandas.pydata.org/docs/getting_started/intro_tutorials/01_table_oriented.html#min-tut-01-tableoriented)?


> A DataFrame is a 2-dimensional data structure that can store data of different types (including characters, integers, floating point values, categorical data and more) in columns.

<img src="https://pandas.pydata.org/docs/_images/01_table_dataframe.svg" title='Pandas DataFrame' width="400" height="200"/>


# Viewing, Describing, and Accessing your Data

Once we have our data as a DataFrame data structure, we can view, describe, and access the data by applying [basic functionalities](https://pandas.pydata.org/pandas-docs/stable/user_guide/basics.html) associated with that structure. For example, we can perform various operations by using the so called `methods`. Examples of `pandas` methods are `head()` and `tail()`; `head()` displays by default the top five rows of the data and `tail()` displays by default the last five rows. You can display a custom number of row by passing that number in the brackets.

## View your data

In [6]:
# Show the first five rows using the method DataFrame.head()
mobility_trends.head()

country_region_code        country_region sub_region_1 sub_region_2  \
0                  AE  United Arab Emirates          NaN          NaN   
1                  AE  United Arab Emirates          NaN          NaN   
2                  AE  United Arab Emirates          NaN          NaN   
3                  AE  United Arab Emirates          NaN          NaN   
4                  AE  United Arab Emirates          NaN          NaN   

  metro_area iso_3166_2_code  census_fips_code                     place_id  \
0        NaN             NaN               NaN  ChIJvRKrsd9IXj4RpwoIwFYv0zM   
1        NaN             NaN               NaN  ChIJvRKrsd9IXj4RpwoIwFYv0zM   
2        NaN             NaN               NaN  ChIJvRKrsd9IXj4RpwoIwFYv0zM   
3        NaN             NaN               NaN  ChIJvRKrsd9IXj4RpwoIwFYv0zM   
4        NaN             NaN               NaN  ChIJvRKrsd9IXj4RpwoIwFYv0zM   

         date  retail_and_recreation_percent_change_from_baseline  \
0  2020-02-15                                                0.0    
1  2020-02-16                                                1.0    
2  2020-02-17                                               -1.0    
3  2020-02-18                                               -2.0    
4  2020-02-19                                               -2.0    

   grocery_and_pharmacy_percent_change_from_baseline  \
0                                                4.0   
1                                                4.0   
2                                                1.0   
3                                                1.0   
4                                                0.0   

   parks_percent_change_from_baseline  \
0                                 5.0   
1                                 4.0   
2                                 5.0   
3                                 5.0   
4                                 4.0   

   transit_stations_percent_change_from_baseline  \
0                                            0.0   
1                                            1.0   
2                                            1.0   
3                                            0.0   
4                                           -1.0   

   workplaces_percent_change_from_baseline  \
0                                      2.0   
1                                      2.0   
2                                      2.0   
3                                      2.0   
4                                      2.0   

   residential_percent_change_from_baseline  
0                                       1.0  
1                                       1.0  
2                                       1.0  
3                                       1.0  
4                                       1.0

In [7]:
# Show the last five rows using the method DataFrame.tail()   
mobility_trends.tail()

country_region_code country_region       sub_region_1 sub_region_2  \
5435339                  ZW       Zimbabwe  Midlands Province       Kwekwe   
5435340                  ZW       Zimbabwe  Midlands Province       Kwekwe   
5435341                  ZW       Zimbabwe  Midlands Province       Kwekwe   
5435342                  ZW       Zimbabwe  Midlands Province       Kwekwe   
5435343                  ZW       Zimbabwe  Midlands Province       Kwekwe   

        metro_area iso_3166_2_code  census_fips_code  \
5435339        NaN             NaN               NaN   
5435340        NaN             NaN               NaN   
5435341        NaN             NaN               NaN   
5435342        NaN             NaN               NaN   
5435343        NaN             NaN               NaN   

                            place_id        date  \
5435339  ChIJRcIZ3-FJNBkRRsj55IcLpfU  2021-05-12   
5435340  ChIJRcIZ3-FJNBkRRsj55IcLpfU  2021-05-13   
5435341  ChIJRcIZ3-FJNBkRRsj55IcLpfU  2021-05-14   
5435342  ChIJRcIZ3-FJNBkRRsj55IcLpfU  2021-05-17   
5435343  ChIJRcIZ3-FJNBkRRsj55IcLpfU  2021-05-18   

         retail_and_recreation_percent_change_from_baseline  \
5435339                                                NaN    
5435340                                                NaN    
5435341                                                NaN    
5435342                                                NaN    
5435343                                                NaN    

         grocery_and_pharmacy_percent_change_from_baseline  \
5435339                                                NaN   
5435340                                                NaN   
5435341                                                NaN   
5435342                                                NaN   
5435343                                                NaN   

         parks_percent_change_from_baseline  \
5435339                                 NaN   
5435340                                 NaN   
5435341                                 NaN   
5435342                                 NaN   
5435343                                 NaN   

         transit_stations_percent_change_from_baseline  \
5435339                                            NaN   
5435340                                            NaN   
5435341                                            NaN   
5435342                                            NaN   
5435343                                            NaN   

         workplaces_percent_change_from_baseline  \
5435339                                     26.0   
5435340                                     19.0   
5435341                                     24.0   
5435342                                     26.0   
5435343                                     35.0   

         residential_percent_change_from_baseline  
5435339                                       NaN  
5435340                                       NaN  
5435341                                       NaN  
5435342                                       NaN  
5435343                                       NaN

In [8]:
# Specify the number of rows to return
mobility_trends.tail(10)

country_region_code country_region       sub_region_1 sub_region_2  \
5435334                  ZW       Zimbabwe  Midlands Province       Kwekwe   
5435335                  ZW       Zimbabwe  Midlands Province       Kwekwe   
5435336                  ZW       Zimbabwe  Midlands Province       Kwekwe   
5435337                  ZW       Zimbabwe  Midlands Province       Kwekwe   
5435338                  ZW       Zimbabwe  Midlands Province       Kwekwe   
5435339                  ZW       Zimbabwe  Midlands Province       Kwekwe   
5435340                  ZW       Zimbabwe  Midlands Province       Kwekwe   
5435341                  ZW       Zimbabwe  Midlands Province       Kwekwe   
5435342                  ZW       Zimbabwe  Midlands Province       Kwekwe   
5435343                  ZW       Zimbabwe  Midlands Province       Kwekwe   

        metro_area iso_3166_2_code  census_fips_code  \
5435334        NaN             NaN               NaN   
5435335        NaN             NaN               NaN   
5435336        NaN             NaN               NaN   
5435337        NaN             NaN               NaN   
5435338        NaN             NaN               NaN   
5435339        NaN             NaN               NaN   
5435340        NaN             NaN               NaN   
5435341        NaN             NaN               NaN   
5435342        NaN             NaN               NaN   
5435343        NaN             NaN               NaN   

                            place_id        date  \
5435334  ChIJRcIZ3-FJNBkRRsj55IcLpfU  2021-05-05   
5435335  ChIJRcIZ3-FJNBkRRsj55IcLpfU  2021-05-06   
5435336  ChIJRcIZ3-FJNBkRRsj55IcLpfU  2021-05-07   
5435337  ChIJRcIZ3-FJNBkRRsj55IcLpfU  2021-05-10   
5435338  ChIJRcIZ3-FJNBkRRsj55IcLpfU  2021-05-11   
5435339  ChIJRcIZ3-FJNBkRRsj55IcLpfU  2021-05-12   
5435340  ChIJRcIZ3-FJNBkRRsj55IcLpfU  2021-05-13   
5435341  ChIJRcIZ3-FJNBkRRsj55IcLpfU  2021-05-14   
5435342  ChIJRcIZ3-FJNBkRRsj55IcLpfU  2021-05-17   
5435343  ChIJRcIZ3-FJNBkRRsj55IcLpfU  2021-05-18   

         retail_and_recreation_percent_change_from_baseline  \
5435334                                                NaN    
5435335                                                NaN    
5435336                                                NaN    
5435337                                                NaN    
5435338                                                NaN    
5435339                                                NaN    
5435340                                                NaN    
5435341                                                NaN    
5435342                                                NaN    
5435343                                                NaN    

         grocery_and_pharmacy_percent_change_from_baseline  \
5435334                                                NaN   
5435335                                                NaN   
5435336                                                NaN   
5435337                                                NaN   
5435338                                                NaN   
5435339                                                NaN   
5435340                                                NaN   
5435341                                                NaN   
5435342                                                NaN   
5435343                                                NaN   

         parks_percent_change_from_baseline  \
5435334                                 NaN   
5435335                                 NaN   
5435336                                 NaN   
5435337                                 NaN   
5435338                                 NaN   
5435339                                 NaN   
5435340                                 NaN   
5435341                                 NaN   
5435342                                 NaN   
5435343                                 NaN   

         transit_stations_percent_change_from_baseline  \
543533

## Describing your DataFrame

In [9]:
# Dimensionality — number of rows and columns — of a DataFrame  
mobility_trends.shape

(5435344, 15)

In [10]:
# Use the print function to display the number of rows and columns in a DataFrame at the time of cell execution
# Import the 'datetime' module to display the current date

import datetime
print("\nAs of", datetime.datetime.now(), "The Google COVID-19 Community Mobility Reports contain", 
      mobility_trends.shape[0], "rows and", mobility_trends.shape[1],"columns.")


As of 2021-05-22 11:47:01.148461 The Google COVID-19 Community Mobility Reports contain 5435344 rows and 15 columns.


In [11]:
# Accessing columns using the DataFrame.columns attribute
mobility_trends.columns

Index(['country_region_code', 'country_region', 'sub_region_1', 'sub_region_2',
       'metro_area', 'iso_3166_2_code', 'census_fips_code', 'place_id', 'date',
       'retail_and_recreation_percent_change_from_baseline',
       'grocery_and_pharmacy_percent_change_from_baseline',
       'parks_percent_change_from_baseline',
       'transit_stations_percent_change_from_baseline',
       'workplaces_percent_change_from_baseline',
       'residential_percent_change_from_baseline'],
      dtype='object')

Let's display a concise summary of our DataFrame using the method `info()`

In [12]:
# Information about a DataFrame
mobility_trends.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5435344 entries, 0 to 5435343
Data columns (total 15 columns):
 #   Column                                              Dtype  
---  ------                                              -----  
 0   country_region_code                                 object 
 1   country_region                                      object 
 2   sub_region_1                                        object 
 3   sub_region_2                                        object 
 4   metro_area                                          object 
 5   iso_3166_2_code                                     object 
 6   census_fips_code                                    float64
 7   place_id                                            object 
 8   date                                                object 
 9   retail_and_recreation_percent_change_from_baseline  float64
 10  grocery_and_pharmacy_percent_change_from_baseline   float64
 11  parks_percent_change_from_baseline   

Among other information, displayed are labels of the columns and the data type of each column. For example 'float64' refers to a floating point number, i.e., a number with a decimal point. In the Community Mobility Reports, floating point numbers are used to display the percentage mobility change for each category of places.

# Accessing columns and rows in your data

## Accessing columns

You can can access columns via column name and column position.

* *Accessing columns via column name*

Let's access a single column.

In [13]:
# Get the country column and save it to its own variable
# The double square bracket option `[[]]` gives DataFrame

mobility_trends_countries = mobility_trends[['country_region']]
mobility_trends_countries.head()

country_region
0  United Arab Emirates
1  United Arab Emirates
2  United Arab Emirates
3  United Arab Emirates
4  United Arab Emirates

In [14]:
# Display the type of data structure

type(mobility_trends_countries)

pandas.core.frame.DataFrame

In [15]:
# The single square braket `[]` option gives Series
# pandas Series is a one-dimensional data structure or simply put, a single column 

mobility_trends['country_region'].head()

0    United Arab Emirates
1    United Arab Emirates
2    United Arab Emirates
3    United Arab Emirates
4    United Arab Emirates
Name: country_region, dtype: object

In [16]:
# Display the type of data structure
type(mobility_trends_countries)

pandas.core.frame.DataFrame

Let's now access multiple columns.

In [17]:
# Accessing multiple columns using a Python list
# You can create lists using square brakets, for example: ['country_region', 'sub_region_1', 'date']
# Python lists are useful as they store multiple items in the same variable 

mobility_trends[['country_region', 'sub_region_1', 'date']].head()

country_region sub_region_1        date
0  United Arab Emirates          NaN  2020-02-15
1  United Arab Emirates          NaN  2020-02-16
2  United Arab Emirates          NaN  2020-02-17
3  United Arab Emirates          NaN  2020-02-18
4  United Arab Emirates          NaN  2020-02-19

* *Accessing columns via column position*

In [18]:
# Accessing columns 1, 2, adn 7 via column position and show the top five rows

mobility_trends.iloc[:, [1, 2, 7]].head()

country_region sub_region_1                     place_id
0  United Arab Emirates          NaN  ChIJvRKrsd9IXj4RpwoIwFYv0zM
1  United Arab Emirates          NaN  ChIJvRKrsd9IXj4RpwoIwFYv0zM
2  United Arab Emirates          NaN  ChIJvRKrsd9IXj4RpwoIwFYv0zM
3  United Arab Emirates          NaN  ChIJvRKrsd9IXj4RpwoIwFYv0zM
4  United Arab Emirates          NaN  ChIJvRKrsd9IXj4RpwoIwFYv0zM

In [19]:
# Accessing a subset of rows and columns. We access the top 3 rows and columns 1, 2, and 7

mobility_trends.iloc[0:3, [1, 2, 7]].head()

country_region sub_region_1                     place_id
0  United Arab Emirates          NaN  ChIJvRKrsd9IXj4RpwoIwFYv0zM
1  United Arab Emirates          NaN  ChIJvRKrsd9IXj4RpwoIwFYv0zM
2  United Arab Emirates          NaN  ChIJvRKrsd9IXj4RpwoIwFYv0zM

# Hands-on exercise

1. Create a new code cell.
2. Access the first 100 rows and the following columns from the DataFrame `mobility_trends`:

    * `country_region_code`
    * `date` 
    * `retail_recreation`

3. Assign the output to a variable called `mobility_trends_subset`.

## Accessing rows

Rows can be accessed via row labels `df.loc` and row index `df.iloc`.

In [20]:
# Before accessing particular rows, let's see the names of all countries in the dataset. We can achieve that by listing all unique values in the 'country_region' column.

mobility_trends.country_region.unique()

array(['United Arab Emirates', 'Afghanistan', 'Antigua and Barbuda',
       'Angola', 'Argentina', 'Austria', 'Australia', 'Aruba',
       'Bosnia and Herzegovina', 'Barbados', 'Bangladesh', 'Belgium',
       'Burkina Faso', 'Bulgaria', 'Bahrain', 'Benin', 'Bolivia',
       'Brazil', 'The Bahamas', 'Botswana', 'Belarus', 'Belize', 'Canada',
       'Switzerland', "Côte d'Ivoire", 'Chile', 'Cameroon', 'Colombia',
       'Costa Rica', 'Cape Verde', 'Czechia', 'Germany', 'Denmark',
       'Dominican Republic', 'Ecuador', 'Estonia', 'Egypt', 'Spain',
       'Finland', 'Fiji', 'France', 'Gabon', 'United Kingdom', 'Georgia',
       'Ghana', 'Greece', 'Guatemala', 'Guinea-Bissau', 'Hong Kong',
       'Honduras', 'Croatia', 'Haiti', 'Hungary', 'Indonesia', 'Ireland',
       'Israel', 'India', 'Iraq', 'Italy', 'Jamaica', 'Jordan', 'Japan',
       'Kenya', 'Kyrgyzstan', 'Cambodia', 'South Korea', 'Kuwait',
       'Kazakhstan', 'Laos', 'Lebanon', 'Liechtenstein', 'Sri Lanka',
       'Lithuania', '

In [65]:
# Accessing specific rows from a DataFrame
# We are interested in the data about the United Kingdom

mobility_trends_UK = mobility_trends[mobility_trends['country_region']=='United Kingdom']
mobility_trends_UK.head()

country_region_code  country_region sub_region_1 sub_region_2  \
1988242                  GB  United Kingdom          NaN          NaN   
1988243                  GB  United Kingdom          NaN          NaN   
1988244                  GB  United Kingdom          NaN          NaN   
1988245                  GB  United Kingdom          NaN          NaN   
1988246                  GB  United Kingdom          NaN          NaN   

        metro_area iso_3166_2_code  census_fips_code  \
1988242        NaN             NaN               NaN   
1988243        NaN             NaN               NaN   
1988244        NaN             NaN               NaN   
1988245        NaN             NaN               NaN   
1988246        NaN             NaN               NaN   

                            place_id        date  \
1988242  ChIJqZHHQhE7WgIReiWIMkOg-MQ  2020-02-15   
1988243  ChIJqZHHQhE7WgIReiWIMkOg-MQ  2020-02-16   
1988244  ChIJqZHHQhE7WgIReiWIMkOg-MQ  2020-02-17   
1988245  ChIJqZHHQhE7WgIReiWIMkOg-MQ  2020-02-18   
1988246  ChIJqZHHQhE7WgIReiWIMkOg-MQ  2020-02-19   

         retail_and_recreation_percent_change_from_baseline  \
1988242                                              -12.0    
1988243                                               -7.0    
1988244                                               10.0    
1988245                                                7.0    
1988246                                                6.0    

         grocery_and_pharmacy_percent_change_from_baseline  \
1988242                                               -7.0   
1988243                                               -6.0   
1988244                                                1.0   
1988245                                               -1.0   
1988246                                               -2.0   

         parks_percent_change_from_baseline  \
1988242                               -35.0   
1988243                               -28.0   
1988244                                24.0   
1988245                                20.0   
1988246                                 8.0   

         transit_stations_percent_change_from_baseline  \
1988242                                          -12.0   
1988243                                           -7.0   
1988244                                           -2.0   
1988245                                           -3.0   
1988246                                           -4.0   

         workplaces_percent_change_from_baseline  \
1988242                                     -4.0   
1988243                                     -3.0   
1988244                                    -14.0   
1988245                                    -14.0   
1988246                                    -14.0   

         residential_percent_change_from_baseline  
1988242                                       2.0  
1988243                                       1.0  
1988244                                       2.0  
1988245                                       2.0  
1988246                                       3.0

In [22]:
# Accessing data about multiple countries using the isin() method
# The backslash symbol \ can be used to split a long command over multiple lines 
mobility_trends_countries = mobility_trends[mobility_trends['country_region'].\
                                                  isin(['United Kingdom', 'Germany', 'Italy', 'Sweden'])]
mobility_trends_countries.head()

country_region_code country_region sub_region_1 sub_region_2  \
1726697                  DE        Germany          NaN          NaN   
1726698                  DE        Germany          NaN          NaN   
1726699                  DE        Germany          NaN          NaN   
1726700                  DE        Germany          NaN          NaN   
1726701                  DE        Germany          NaN          NaN   

        metro_area iso_3166_2_code  census_fips_code  \
1726697        NaN             NaN               NaN   
1726698        NaN             NaN               NaN   
1726699        NaN             NaN               NaN   
1726700        NaN             NaN               NaN   
1726701        NaN             NaN               NaN   

                            place_id        date  \
1726697  ChIJa76xwh5ymkcRW-WRjmtd6HU  2020-02-15   
1726698  ChIJa76xwh5ymkcRW-WRjmtd6HU  2020-02-16   
1726699  ChIJa76xwh5ymkcRW-WRjmtd6HU  2020-02-17   
1726700  ChIJa76xwh5ymkcRW-WRjmtd6HU  2020-02-18   
1726701  ChIJa76xwh5ymkcRW-WRjmtd6HU  2020-02-19   

         retail_and_recreation_percent_change_from_baseline  \
1726697                                                6.0    
1726698                                                7.0    
1726699                                                2.0    
1726700                                                2.0    
1726701                                                3.0    

         grocery_and_pharmacy_percent_change_from_baseline  \
1726697                                                1.0   
1726698                                               10.0   
1726699                                                2.0   
1726700                                                2.0   
1726701                                                0.0   

         parks_percent_change_from_baseline  \
1726697                                45.0   
1726698                                 9.0   
1726699                                 7.0   
1726700                                10.0   
1726701                                 6.0   

         transit_stations_percent_change_from_baseline  \
1726697                                           10.0   
1726698                                            6.0   
1726699                                            1.0   
1726700                                            1.0   
1726701                                           -1.0   

         workplaces_percent_change_from_baseline  \
1726697                                      0.0   
1726698                                     -1.0   
1726699                                     -2.0   
1726700                                     -1.0   
1726701                                     -1.0   

         residential_percent_change_from_baseline  
1726697                                      -1.0  
1726698                                       0.0  
1726699                                       0.0  
1726700                                       1.0  
1726701                                       1.0

In [23]:
# Filter by two conditions — country and county — simultenously
# First let's see the list of counties in the dataset

mobility_trends[mobility_trends['country_region']=='United Kingdom'].sub_region_1.unique()

array([nan, 'Aberdeen City', 'Aberdeenshire', 'Angus Council',
       'Antrim and Newtownabbey', 'Ards and North Down',
       'Argyll and Bute Council', 'Armagh City, Banbridge and Craigavon',
       'Bath and North East Somerset', 'Bedford', 'Belfast',
       'Blackburn with Darwen', 'Blackpool', 'Blaenau Gwent',
       'Borough of Halton', 'Bracknell Forest', 'Bridgend County Borough',
       'Brighton and Hove', 'Bristol City', 'Buckinghamshire',
       'Caerphilly County Borough', 'Cambridgeshire', 'Cardiff',
       'Carmarthenshire', 'Causeway Coast and Glens',
       'Central Bedfordshire', 'Ceredigion', 'Cheshire East',
       'Cheshire West and Chester', 'Clackmannanshire',
       'Conwy Principal Area', 'Cornwall', 'County Durham', 'Cumbria',
       'Darlington', 'Denbighshire', 'Derby', 'Derbyshire',
       'Derry and Strabane', 'Devon', 'Dorset', 'Dumfries and Galloway',
       'Dundee City Council', 'East Ayrshire Council',
       'East Dunbartonshire Council', 'East Lothi

In [24]:
# Access data about UK and Essex
mobility_trends_UK_Essex = mobility_trends[(mobility_trends['country_region'] == 'United Kingdom') & 
                                 (mobility_trends['sub_region_1']=='Essex')]

mobility_trends_UK_Essex.head()

country_region_code  country_region sub_region_1 sub_region_2  \
2028787                  GB  United Kingdom        Essex          NaN   
2028788                  GB  United Kingdom        Essex          NaN   
2028789                  GB  United Kingdom        Essex          NaN   
2028790                  GB  United Kingdom        Essex          NaN   
2028791                  GB  United Kingdom        Essex          NaN   

        metro_area iso_3166_2_code  census_fips_code  \
2028787        NaN          GB-ESS               NaN   
2028788        NaN          GB-ESS               NaN   
2028789        NaN          GB-ESS               NaN   
2028790        NaN          GB-ESS               NaN   
2028791        NaN          GB-ESS               NaN   

                            place_id        date  \
2028787  ChIJ0w2H_idW2EcReVDuRzjLV0I  2020-02-15   
2028788  ChIJ0w2H_idW2EcReVDuRzjLV0I  2020-02-16   
2028789  ChIJ0w2H_idW2EcReVDuRzjLV0I  2020-02-17   
2028790  ChIJ0w2H_idW2EcReVDuRzjLV0I  2020-02-18   
2028791  ChIJ0w2H_idW2EcReVDuRzjLV0I  2020-02-19   

         retail_and_recreation_percent_change_from_baseline  \
2028787                                               -9.0    
2028788                                              -16.0    
2028789                                                9.0    
2028790                                               10.0    
2028791                                                9.0    

         grocery_and_pharmacy_percent_change_from_baseline  \
2028787                                               -7.0   
2028788                                              -12.0   
2028789                                               -1.0   
2028790                                               -4.0   
2028791                                               -3.0   

         parks_percent_change_from_baseline  \
2028787                               -27.0   
2028788                               -45.0   
2028789                                23.0   
2028790                                15.0   
2028791                                10.0   

         transit_stations_percent_change_from_baseline  \
2028787                                          -11.0   
2028788                                           -9.0   
2028789                                           -1.0   
2028790                                            0.0   
2028791                                            1.0   

         workplaces_percent_change_from_baseline  \
2028787                                     -3.0   
2028788                                     -5.0   
2028789                                    -16.0   
2028790                                    -16.0   
2028791                                    -16.0   

         residential_percent_change_from_baseline  
2028787                                       2.0  
2028788                                       2.0  
2028789                                       3.0  
2028790                                       2.0  
2028791                                       2.0

In the above example, you could simply type in  
```
mobility_trends[mobility_trends['sub_region_1']=='Essex']
```
The command will return the same result as there is no region called 'Essex' outside the UK. However, there are instances in the data where regions in different countries have the same name. When this is the case, not specifying the country would return confusing outputs.

## Accessing multiple rows and columns and conditioning

In [25]:
mobility_trends_UK

country_region_code  country_region sub_region_1 sub_region_2  \
1988242                  GB  United Kingdom          NaN          NaN   
1988243                  GB  United Kingdom          NaN          NaN   
1988244                  GB  United Kingdom          NaN          NaN   
1988245                  GB  United Kingdom          NaN          NaN   
1988246                  GB  United Kingdom          NaN          NaN   
...                     ...             ...          ...          ...   
2179301                  GB  United Kingdom         York          NaN   
2179302                  GB  United Kingdom         York          NaN   
2179303                  GB  United Kingdom         York          NaN   
2179304                  GB  United Kingdom         York          NaN   
2179305                  GB  United Kingdom         York          NaN   

        metro_area iso_3166_2_code  census_fips_code  \
1988242        NaN             NaN               NaN   
1988243        NaN             NaN               NaN   
1988244        NaN             NaN               NaN   
1988245        NaN             NaN               NaN   
1988246        NaN             NaN               NaN   
...            ...             ...               ...   
2179301        NaN          GB-YOR               NaN   
2179302        NaN          GB-YOR               NaN   
2179303        NaN          GB-YOR               NaN   
2179304        NaN          GB-YOR               NaN   
2179305        NaN          GB-YOR               NaN   

                            place_id        date  \
1988242  ChIJqZHHQhE7WgIReiWIMkOg-MQ  2020-02-15   
1988243  ChIJqZHHQhE7WgIReiWIMkOg-MQ  2020-02-16   
1988244  ChIJqZHHQhE7WgIReiWIMkOg-MQ  2020-02-17   
1988245  ChIJqZHHQhE7WgIReiWIMkOg-MQ  2020-02-18   
1988246  ChIJqZHHQhE7WgIReiWIMkOg-MQ  2020-02-19   
...                              ...         ...   
2179301  ChIJh-IigLwxeUgRAKFv7Z75DAM  2021-05-14   
2179302  ChIJh-IigLwxeUgRAKFv7Z75DAM  2021-05-15   
2179303  ChIJh-IigLwxeUgRAKFv7Z75DAM  2021-05-16   
2179304  ChIJh-IigLwxeUgRAKFv7Z75DAM  2021-05-17   
2179305  ChIJh-IigLwxeUgRAKFv7Z75DAM  2021-05-18   

         retail_and_recreation_percent_change_from_baseline  \
1988242                                              -12.0    
1988243                                               -7.0    
1988244                                               10.0    
1988245                                                7.0    
1988246                                                6.0    
...                                                    ...    
2179301                                              -33.0    
2179302                                              -40.0    
2179303                                              -36.0    
2179304                                              -14.0    
2179305                                              -11.0    

         grocery_and_pharmacy_percent_change_from_baseline  \
1988242                                               -7.0   
1988243                                               -6.0   
1988244                                                1.0   
1988245                                               -1.0   
1988246                                               -2.0   
...                                                    ...   
2179301                                               10.0   
2179302                                                4.0   
2179303                                                4.0   
2179304                                                7.0   
2179305                                                9.0   

         parks_percent_change_from_baseline  \
1988242                               -35.0   
1988243                               -28.0   
1988244                                24.0   
1988245                                20.0   
1988246                                 8.0   
...                                     ...   
21

In [26]:
# Let's see which UK counties had the lower retail and recreation mobility the day after Italy went in lockdown on 9 March 2020
mobility_trends_UK_10MARCH2020 = mobility_trends_UK.loc[(mobility_trends_UK['date']=='2020-03-10') &
                                       (mobility_trends_UK['retail_and_recreation_percent_change_from_baseline']<0),
                                       ['sub_region_1','retail_and_recreation_percent_change_from_baseline']]

# Sort in decreasing order
mobility_trends_UK_10MARCH2020.sort_values(by='retail_and_recreation_percent_change_from_baseline', ascending=True)

sub_region_1  \
2039322     Gloucestershire   
2040690      Greater London   
2053083      Greater London   
2041149      Greater London   
2050329      Greater London   
...                     ...   
2155980       Tyne and Wear   
2120655     Nottinghamshire   
2058132  Greater Manchester   
2145929             Suffolk   
2075038                Kent   

         retail_and_recreation_percent_change_from_baseline  
2039322                                              -16.0   
2040690                                              -14.0   
2053083                                              -13.0   
2041149                                              -12.0   
2050329                                              -11.0   
...                                                    ...   
2155980                                               -1.0   
2120655                                               -1.0   
2058132                                               -1.0   
2145929                                               -1.0   
2075038                                               -1.0   

[167 rows x 2 columns]

In [27]:
# UK counties with the lower retail and recreation the day after UK went in lockdown on 23 March 2020 
mobility_trends_UK_24MARCH2020=mobility_trends_UK.loc[(mobility_trends_UK['date']=='2020-03-24') & 
                                       (mobility_trends_UK['retail_and_recreation_percent_change_from_baseline']<0), 
                                       ['sub_region_1','retail_and_recreation_percent_change_from_baseline']]

# Sort in decreasing order
mobility_trends_UK_24MARCH2020.sort_values(by='retail_and_recreation_percent_change_from_baseline', ascending=True)

sub_region_1  retail_and_recreation_percent_change_from_baseline
2040704  Greater London                                              -94.0 
2041163  Greater London                                              -92.0 
2043917  Greater London                                              -87.0 
2054933  Greater London                                              -84.0 
2048966  Greater London                                              -83.0 
...                 ...                                                ... 
2167904     West Sussex                                              -56.0 
2010094         Cumbria                                              -56.0 
2077806            Kent                                              -54.0 
2078724            Kent                                              -54.0 
2097026      Merseyside                                              -50.0 

[417 rows x 2 columns]

Let's identify the places in the UK with the highest reduction of workplaces mobility.

In [28]:
# We need to first find the index of the row with the highest reduction of workplaces mobility
mobility_trends_UK_workplaces = mobility_trends_UK['workplaces_percent_change_from_baseline'].idxmin()
mobility_trends_UK_workplaces

2048986

In [29]:
# Use the above index to access the row with the highest reduction of workplaces mobility
mobility_trends.iloc[[mobility_trends_UK_workplaces]]

country_region_code  country_region    sub_region_1  \
2048986                  GB  United Kingdom  Greater London   

                        sub_region_2 metro_area iso_3166_2_code  \
2048986  London Borough of Islington        NaN          GB-ISL   

         census_fips_code                     place_id        date  \
2048986               NaN  ChIJ5Tvr7V0bdkgRyWXsGMuF8FQ  2020-04-13   

         retail_and_recreation_percent_change_from_baseline  \
2048986                                              -88.0    

         grocery_and_pharmacy_percent_change_from_baseline  \
2048986                                              -51.0   

         parks_percent_change_from_baseline  \
2048986                               -28.0   

         transit_stations_percent_change_from_baseline  \
2048986                                          -90.0   

         workplaces_percent_change_from_baseline  \
2048986                                    -91.0   

         residential_percent_change_from_baseline  
2048986                                      40.0

In our basic exploratory data analysis, we found that as of 21 April 2021 the largest percentage mobility change in workplaces was in the London Borough of Islington, Greater London, on 13 April 2020 when the workplaces mobility was reduced with 91 per cent compared to a baseline period. This result is only for exploratory purposes. Note that [the same mobility in two different days would result in different percentage changes](https://support.google.com/covid19-mobility/answer/9824897?hl=en&ref_topic=9822927) due to the way in which actual mobility is compared to a baseline in the Google Covid-19 Community Mobility Reports.

# Handling Missing Values
Many data sets have missing values and the mobility data is not an exception. Handling missing values is a very common problem in data analysis. We will cover two simple ways of handling missing values in data: 
* Deletion — missing observations are simply dropped from the data set. 
    * Likewise - removes all data for an observation that has one or more missing values.     
    * Pairwise — removes observations with missing values only for the variables used in the current procedure (e.g., model or technique).
* Imputation — missing observations are replaced with inferred values. We will cover simple imputation methods in which missing values are replaced with column mean or row mean.

For more advanced methods and functionality, including interpolation methods, see the section ['Working with missing data'](https://pandas.pydata.org/pandas-docs/version/0.19.1/missing_data.html) in Pandas documentation.

## Detecting missing values
In `pandas`, missing values are represented as `NaN`, which stands for Not a Number. To detect missing values, we can use the function `isna()` (other functions, for example `isnull()` can be used to the same effect). The function returns boolean values (False, True) indicating whether each corresponding value in a DataFrame is missing.

In [71]:
mobility_trends.isna()

country_region_code  country_region  sub_region_1  sub_region_2  \
0                      False           False          True          True   
1                      False           False          True          True   
2                      False           False          True          True   
3                      False           False          True          True   
4                      False           False          True          True   
...                      ...             ...           ...           ...   
5461522                False           False         False         False   
5461523                False           False         False         False   
5461524                False           False         False         False   
5461525                False           False         False         False   
5461526                False           False         False         False   

         metro_area  iso_3166_2_code  census_fips_code  place_id   date  \
0              True             True              True     False  False   
1              True             True              True     False  False   
2              True             True              True     False  False   
3              True             True              True     False  False   
4              True             True              True     False  False   
...             ...              ...               ...       ...    ...   
5461522        True             True              True     False  False   
5461523        True             True              True     False  False   
5461524        True             True              True     False  False   
5461525        True             True              True     False  False   
5461526        True             True              True     False  False   

         retail_and_recreation_percent_change_from_baseline  \
0                                                    False    
1                                                    False    
2                                                    False    
3                                                    False    
4                                                    False    
...                                                    ...    
5461522                                               True    
5461523                                               True    
5461524                                               True    
5461525                                               True    
5461526                                               True    

         grocery_and_pharmacy_percent_change_from_baseline  \
0                                                    False   
1                                                    False   
2                                                    False   
3                                                    False   
4                                                    False   
...                                                    ...   
5461522                                               True   
5461523                                               True   
5461524                                               True   
5461525                                               True   
5461526                                               True   

         parks_percent_change_from_baseline  \
0                                     False   
1                                     False   
2                                     False   
3                                     False   
4                                     False   
...                                     ...   
5461522                                True   
5461523                                True   
5461524                                True   
5461525                                True   
5461526                                True   

         transit_stations_percent_change_from_baseline  \
0                                                False   
1           

Let's determine the number of missing values for each column in the DataFrame.

In [72]:
mobility_trends.isna().sum()

country_region_code                                      3479
country_region                                              0
sub_region_1                                            91843
sub_region_2                                           900937
metro_area                                            5431697
iso_3166_2_code                                       4492972
census_fips_code                                      4307968
place_id                                                 9879
date                                                        0
retail_and_recreation_percent_change_from_baseline    2052299
grocery_and_pharmacy_percent_change_from_baseline     2183223
parks_percent_change_from_baseline                    2839762
transit_stations_percent_change_from_baseline         2715596
workplaces_percent_change_from_baseline                227198
residential_percent_change_from_baseline              2234460
dtype: int64

It would be useful to quantify the percentage of values that are missing per column. To achieve this, we simply divide the number of missing values in a column by the number of rows in the DataFrame (and multiply the result by 100).

In [73]:
mobility_trends.isna().sum()/mobility_trends.shape[0]*100

country_region_code                                    0.063700
country_region                                         0.000000
sub_region_1                                           1.681636
sub_region_2                                          16.496064
metro_area                                            99.453816
iso_3166_2_code                                       82.265857
census_fips_code                                      78.878453
place_id                                               0.180883
date                                                   0.000000
retail_and_recreation_percent_change_from_baseline    37.577384
grocery_and_pharmacy_percent_change_from_baseline     39.974590
parks_percent_change_from_baseline                    51.995751
transit_stations_percent_change_from_baseline         49.722285
workplaces_percent_change_from_baseline                4.159972
residential_percent_change_from_baseline              40.912734
dtype: float64

Let's now repeat the procedure for the UK mobility data only: 

In [74]:
mobility_trends_UK.isna().sum()/mobility_trends_UK.shape[0]*100

country_region_code                                     0.000000
country_region                                          0.000000
sub_region_1                                            0.240234
sub_region_2                                           36.080580
metro_area                                            100.000000
iso_3166_2_code                                        49.205502
census_fips_code                                      100.000000
place_id                                                0.000000
date                                                    0.000000
retail_and_recreation_percent_change_from_baseline      3.949462
grocery_and_pharmacy_percent_change_from_baseline       3.523950
parks_percent_change_from_baseline                     21.737742
transit_stations_percent_change_from_baseline           4.204874
workplaces_percent_change_from_baseline                 0.846837
residential_percent_change_from_baseline                8.076875
dtype: float64

The output shows that the missing values in most mobility categories are below 5 per cent except the missing values for residential mobility (8%) and parks mobility (21.7%).

## Removing missing values

We can remove missing values using the Pandas function [`dropna()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dropna.html). The default functionality removes rows that contain at least one missing value. The `dropna()` function provides other capabilities as well. For example, it is possible to remove columns (instead of rows) that contain missing values by specifying `axis = 1` within the brackets. Also, we can specify under what conditions rows or columns are removed: use `how = any` if you would like to drop rows or columns that contain any missing value, which is the default functionality; use `how = all` if you would like to drop rows or columns in which all values are missing.

In [60]:
mobility_trends_UK.dropna()

Empty DataFrame
Columns: [country_region_code, country_region, sub_region_1, sub_region_2, metro_area, iso_3166_2_code, census_fips_code, place_id, date, retail_and_recreation_percent_change_from_baseline, grocery_and_pharmacy_percent_change_from_baseline, parks_percent_change_from_baseline, transit_stations_percent_change_from_baseline, workplaces_percent_change_from_baseline, residential_percent_change_from_baseline]
Index: []

We did not specify particular columns on which to perform the `dropna()` function. Because we have a column ('metro_area') for which all values are missing, the `dropna()` function removed all observations from the DataFrame. This is analogous to a likewise deletion of missing values.

A more informative approach would be to specify the columns we are going to use in our analysis and drop missing values only for this subset of variables. We can drop rows that contain any missing values in the six mobility categories by passing the labels of those categories to the parameter `subset` as shown below:

In [62]:
mobility_trends_UK.dropna(subset=mobility_trends_UK.iloc[:,9:15].columns)

country_region_code  country_region sub_region_1 sub_region_2  \
1988242                  GB  United Kingdom          NaN          NaN   
1988243                  GB  United Kingdom          NaN          NaN   
1988244                  GB  United Kingdom          NaN          NaN   
1988245                  GB  United Kingdom          NaN          NaN   
1988246                  GB  United Kingdom          NaN          NaN   
...                     ...             ...          ...          ...   
2179301                  GB  United Kingdom         York          NaN   
2179302                  GB  United Kingdom         York          NaN   
2179303                  GB  United Kingdom         York          NaN   
2179304                  GB  United Kingdom         York          NaN   
2179305                  GB  United Kingdom         York          NaN   

        metro_area iso_3166_2_code  census_fips_code  \
1988242        NaN             NaN               NaN   
1988243        NaN             NaN               NaN   
1988244        NaN             NaN               NaN   
1988245        NaN             NaN               NaN   
1988246        NaN             NaN               NaN   
...            ...             ...               ...   
2179301        NaN          GB-YOR               NaN   
2179302        NaN          GB-YOR               NaN   
2179303        NaN          GB-YOR               NaN   
2179304        NaN          GB-YOR               NaN   
2179305        NaN          GB-YOR               NaN   

                            place_id        date  \
1988242  ChIJqZHHQhE7WgIReiWIMkOg-MQ  2020-02-15   
1988243  ChIJqZHHQhE7WgIReiWIMkOg-MQ  2020-02-16   
1988244  ChIJqZHHQhE7WgIReiWIMkOg-MQ  2020-02-17   
1988245  ChIJqZHHQhE7WgIReiWIMkOg-MQ  2020-02-18   
1988246  ChIJqZHHQhE7WgIReiWIMkOg-MQ  2020-02-19   
...                              ...         ...   
2179301  ChIJh-IigLwxeUgRAKFv7Z75DAM  2021-05-14   
2179302  ChIJh-IigLwxeUgRAKFv7Z75DAM  2021-05-15   
2179303  ChIJh-IigLwxeUgRAKFv7Z75DAM  2021-05-16   
2179304  ChIJh-IigLwxeUgRAKFv7Z75DAM  2021-05-17   
2179305  ChIJh-IigLwxeUgRAKFv7Z75DAM  2021-05-18   

         retail_and_recreation_percent_change_from_baseline  \
1988242                                              -12.0    
1988243                                               -7.0    
1988244                                               10.0    
1988245                                                7.0    
1988246                                                6.0    
...                                                    ...    
2179301                                              -33.0    
2179302                                              -40.0    
2179303                                              -36.0    
2179304                                              -14.0    
2179305                                              -11.0    

         grocery_and_pharmacy_percent_change_from_baseline  \
1988242                                               -7.0   
1988243                                               -6.0   
1988244                                                1.0   
1988245                                               -1.0   
1988246                                               -2.0   
...                                                    ...   
2179301                                               10.0   
2179302                                                4.0   
2179303                                                4.0   
2179304                                                7.0   
2179305                                                9.0   

         parks_percent_change_from_baseline  \
1988242                               -35.0   
1988243                               -28.0   
1988244                                24.0   
1988245                                20.0   
1988246                                 8.0   
...                                     ...   
21

In [50]:
mobility_trends_UK.shape

(191064, 15)

After removing the rows that contain missing values in the six mobility categories/columns, we are left with 137992 rows out of 191064 rows, a reduction of approximately 28%.

## Replacing missing values

Let’s replace NaN values with the mean of non-missing values. For example, we can replace missing values in the columns 'workplaces_percent_change_from_baseline' and 'residential_percent_change_from_baseline' by the mean of values in the two columns as determined by the Pandas `mean()` method.

In [67]:
# Replace the NaNs in columns 'workplaces_percent_change_from_baseline' and 'residential_percent_change_from_baseline' by the mean of non-missing values
mobility_trends_UK[['workplaces_percent_change_from_baseline','residential_percent_change_from_baseline']].\
fillna(value=mobility_trends_UK[['workplaces_percent_change_from_baseline','residential_percent_change_from_baseline']].mean())

workplaces_percent_change_from_baseline  \
1988242                                     -4.0   
1988243                                     -3.0   
1988244                                    -14.0   
1988245                                    -14.0   
1988246                                    -14.0   
...                                          ...   
2179301                                    -38.0   
2179302                                    -22.0   
2179303                                    -21.0   
2179304                                    -38.0   
2179305                                    -38.0   

         residential_percent_change_from_baseline  
1988242                                       2.0  
1988243                                       1.0  
1988244                                       2.0  
1988245                                       2.0  
1988246                                       3.0  
...                                           ...  
2179301                                      11.0  
2179302                                       7.0  
2179303                                       6.0  
2179304                                      10.0  
2179305                                      10.0  

[191064 rows x 2 columns]

We now have the same number of rows (191064) as the original mobility_trends_UK DataFrame, with each missing value replaced by the mean value for the respective column. More advanced (and practical) approaches of replacing missing values can take into account groupings in the data and replace missing values with the mean value of a group of counties or the mean value of a time period; those operations would require more advanced Pandas functionality such as `groupby()`, which we will use in the next lab.

<img src="https://covid19-static.cdn-apple.com/applications/covid19-mobility/current/static/og_image_v1.png" width="200" height="100" align="right"/>

# Compare mobility data formats

A related aggregate, privacy-preserving, and freely available mobility data set is the Apple's [Mobility Trends Reports](https://covid19.apple.com/mobility). The data is updated daily and reflect in aggregate requests for directions in Apple Maps for select countries/regions, sub-regions and cities.  

The Mobility Trends Reports data can be downloaded from the website https://covid19.apple.com/mobility. However, because the data file is not at a stable location (URL), we cannot load the data from the website but instead access the data from a [COVID-19 mobility data aggregator on GitHub](https://github.com/ActiveConclusion/COVID19_mobility).

In [ ]:
mobility_trends_covid_apple = pd.read_csv('https://raw.githubusercontent.com/ActiveConclusion/COVID19_mobility/master/apple_reports/applemobilitytrends.csv')
mobility_trends_covid_apple.head()

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (3) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


geo_type     region  ... 2021-04-19 2021-04-20
0  country/region    Albania  ...     116.15     114.33
1  country/region    Albania  ...     149.68     142.30
2  country/region  Argentina  ...      60.19      63.25
3  country/region  Argentina  ...      56.19      58.81
4  country/region  Australia  ...     109.87     113.70

[5 rows x 470 columns]

```{admonition} Activity: Discussing data formats
Compare the format of the data in the _Apple Mobility Trends Reports_ to the format of the data in the _Google COVID-19 Community Mobility Reports_ and in _Our World in Data COVID-19 data_. Focus in particular on similarities and differences in how time series data is represented across data sets.
```

# Recording dependencies

In [142]:
# Install the watermark extension
!pip install watermark

# Load the watermark extension
%load_ext watermark

# Show packages that were imported
%watermark --iversions

The watermark extension is already loaded. To reload it, use:
  %reload_ext watermark
matplotlib: 3.3.2
pandas    : 1.1.3
seaborn   : 0.11.0

